internal/sessions/cache/cache_store.go

```diff
@@ -1,98 +1,82 @@
 package cache // import "github.com/pomerium/pomerium/internal/sessions/cache"
 
 import (
-	"context"
 	"errors"
 	"fmt"
 	"net/http"
 
-	"github.com/golang/groupcache"
-
 	"github.com/pomerium/pomerium/internal/encoding"
+	"github.com/pomerium/pomerium/internal/grpc/cache/client"
 	"github.com/pomerium/pomerium/internal/log"
 	"github.com/pomerium/pomerium/internal/sessions"
 )
 
 var _ sessions.SessionStore = &Store{}
 var _ sessions.SessionLoader = &Store{}
 
-const (
-	defaultQueryParamKey = "ati"
-)
-
-// Store implements the session store interface using a distributed cache.
+// Store implements the session store interface using a cache service.
 type Store struct {
-	name    string
-	encoder encoding.Marshaler
-	decoder encoding.Unmarshaler
-
-	cache        *groupcache.Group
+	cache        client.Cacher
+	encoder      encoding.MarshalUnmarshaler
+	queryParam   string
 	wrappedStore sessions.SessionStore
 }
 
-// defaultCacheSize is ~10MB
-var defaultCacheSize int64 = 10 << 20
-
-// NewStore creates a new session store built on the distributed caching library
-// groupcache. On a cache miss, the cache store attempts to fallback to another
-// SessionStore implementation.
-func NewStore(enc encoding.MarshalUnmarshaler, wrappedStore sessions.SessionStore, name string) *Store {
-	store := &Store{
-		name:         name,
-		encoder:      enc,
-		decoder:      enc,
-		wrappedStore: wrappedStore,
-	}
+// Options represent cache store's available configurations.
+type Options struct {
+	Cache        client.Cacher
+	Encoder      encoding.MarshalUnmarshaler
+	QueryParam   string
+	WrappedStore sessions.SessionStore
+}
 
-	store.cache = groupcache.NewGroup(name, defaultCacheSize, groupcache.GetterFunc(
-		func(ctx context.Context, id string, dest groupcache.Sink) error {
-			// fill the cache with session set as part of the request
-			// context set previously as part of SaveSession.
-			b := fromContext(ctx)
-			if len(b) == 0 {
-				return fmt.Errorf("sessions/cache: cannot fill key %s from ctx", id)
-			}
-			if err := dest.SetBytes(b); err != nil {
-				return fmt.Errorf("sessions/cache: sink error %w", err)
-			}
-			return nil
-		},
-	))
+var defaultOptions = &Options{
+	QueryParam: "ati",
```

Comment:
https://github.com/pomerium/pomerium/blob/8a9cb0f803ac7d1b4c28d97956588f242f4d612e/internal/sessions/state.go#L57

https://github.com/pomerium/pomerium/blob/8a9cb0f803ac7d1b4c28d97956588f242f4d612e/proxy/middleware.go#L65

in the future, I will refactor to make the default something more generic, I think.

Reply:
👍

Reply:
Actually, maybe that should be a const in `pomerium/internal/sessions/state.go` since it's spread out in 3 files.  Not sure if that would cause dependency tree issues, though.

Reply:
Good call.